### PR TITLE
[Issue #371] Add NetNTLMv2 hash capture to README

### PR DESCRIPTION
**Pull request by lubiedo**
@obilodeau Let me know if there's more to be added

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -348,6 +348,16 @@ For example, if you expect your payload to take up to 5 seconds to complete, you
 This will block the client's input / output for 5 seconds to hide the console and prevent interference.
 After 5 seconds, input / output is restored back to normal.
 
+#### Capturing NetNTLMv2 hashes
+PyRDP has the ability to capture the client's NetNTLMv2 hashes via a NLA (CredSSP)
+connection by carrying the negotiation and capturing the NTLMSSP authentication
+messages. This will happen even if the connection against the RDP remote server
+is not done via NLA as PyRDP will generate NTLMSSP messages by it's own and
+continue with the authentication process. The capturing NetNTLMv2 hashes can be
+useful for research and offensivepurposes.
```
Comment (obilodeau) [inline README.md:357]:
```suggestion
useful for research and offensive purposes.
```

File: README.md [hunk 2]
```diff
@@ -348,6 +348,16 @@ For example, if you expect your payload to take up to 5 seconds to complete, you
 This will block the client's input / output for 5 seconds to hide the console and prevent interference.
 After 5 seconds, input / output is restored back to normal.
 
+#### Capturing NetNTLMv2 hashes
+PyRDP has the ability to capture the client's NetNTLMv2 hashes via a NLA (CredSSP)
+connection by carrying the negotiation and capturing the NTLMSSP authentication
+messages. This will happen even if the connection against the RDP remote server
+is not done via NLA as PyRDP will generate NTLMSSP messages by it's own and
```
Comment (obilodeau) [inline README.md:355]: This is inaccurate in non-NLA contexts. I should be further clarified.

File: README.md [hunk 3]
```diff
@@ -348,6 +348,16 @@ For example, if you expect your payload to take up to 5 seconds to complete, you
...
+continue with the authentication process. The capturing NetNTLMv2 hashes can be
+useful for research and offensivepurposes.
+The captured NetNTLMv2 hash can be found in the logs and it's formatted so
```
Comment (obilodeau) [inline README.md:358]: say which log

**Review by obilodeau — CHANGES_REQUESTED** [review]
This addresses one portion of the request in https://github.com/GoSecure/pyrdp/issues/371. The other two need to be done as well.

**Comment by lubiedo** [comment]
@obilodeau Committed some changes, let me know if everything is OK

File: README.md [hunk 4]
```diff
@@ -276,6 +276,22 @@ If key generation didn't work or you want to use a custom key and certificate, y
 pyrdp-mitm.py 192.168.1.10 -k private_key.pem -c certificate.pem
 ```
 
+##### Monster-in-the-Middle NLA/CredSSP connections
+You can also specify the server's key and certificate along with the
+authentication as `ssp` to start an NLA connection against the server:
+```
+--auth ssp
+```
+This will enable the possibility to intercept NLA enabled connections.
```
Comment (obilodeau) [inline README.md:285]:
```suggestion
This will enable the possibility to intercept NLA-enforced connections.
```

File: README.md [hunk 5]
```diff
@@ -276,6 +276,22 @@ If key generation didn't work or you want to use a custom key and certificate, y
 pyrdp-mitm.py 192.168.1.10 -k private_key.pem -c certificate.pem
 ```
 
+##### Monster-in-the-Middle NLA/CredSSP connections
+You can also specify the server's key and certificate along with the
```
Comment (obilodeau) [inline README.md:280]: TODO
- [ ] link to blogpost introducing the feature
- [ ] link to cert extraction documentation in `docs/`

File: README.md [hunk 6]
```diff
@@ -276,6 +276,22 @@ If key generation didn't work or you want to use a custom key and certificate, y
 pyrdp-mitm.py 192.168.1.10 -k private_key.pem -c certificate.pem
 ```
 
+##### Monster-in-the-Middle NLA/CredSSP connections
```
Comment (obilodeau) [inline README.md:279]: Expand NLA to Network Level Authentication (NLA) the first time it is used

File: README.md [hunk 7]
```diff
@@ -276,6 +276,22 @@ If key generation didn't work or you want to use a custom key and certificate, y
...
+You can also specify the server's key and certificate along with the
+authentication as `ssp` to start an NLA connection against the server:
+```
+--auth ssp
+```
+This will enable the possibility to intercept NLA enabled connections.
+
+##### Alternative host redirection
+If the server you are trying to connect enforces NLA, you can redirect the
+connection to another non-NLA server to keep the connection alive. To enable
```
Comment (obilodeau) [inline README.md:289]: TODO
- [ ] more context about when to use this attack

File: README.md [hunk 8]
```diff
@@ -276,6 +276,22 @@ If key generation didn't work or you want to use a custom key and certificate, y
...
+this feature append specify the alternative host's address and port:
+```
+--nla-redirection-host 192.168.1.12 --nla-redirection-port 3389
+```
```
Comment (obilodeau) [inline README.md:293]: Add a diagram, I will also use it in the blog post since this was added in 1.1.0 which is covered by my next blog

File: README.md [hunk 9]
```diff
@@ -348,6 +364,16 @@ For example, if you expect your payload to take up to 5 seconds to complete, you
 This will block the client's input / output for 5 seconds to hide the console and prevent interference.
 After 5 seconds, input / output is restored back to normal.
 
+#### Capturing NetNTLMv2 hashes
```
Comment (obilodeau) [inline README.md:367]: TODO: Refer to Responder for people looking for a simpler tool

File: README.md [hunk 10]
```diff
@@ -348,6 +364,16 @@ For example, if you expect your payload to take up to 5 seconds to complete, you
...
+PyRDP has the ability to capture the client's NetNTLMv2 hashes via a NLA
+(CredSSP) connection by carrying the negotiation and capturing the NTLMSSP
+authentication messages. The capturing NetNTLMv2 hashes can be useful for
+research and offensive purposes. As last resort, PyRDP will also try to capture
+hashes from connections where the server enforces NLA but PyRDP was initiated
+as non-NLA (`tls`).
```
Comment (obilodeau) [inline README.md:373]: reword this to avoid talking about PyRDP's NLA/non-NLA (confusing)

**Review by obilodeau — COMMENTED** [review]
Did another round of review. The notes are mostly to myself, I think I should pick-it up from here.

**Comment by obilodeau** [comment]
Dammit I screwed up, I brought in all changes in master since PR was created by accident... Review can focus on my last commit.

**Review by obilodeau — APPROVED** [review]
I finished the changes
